tests/sentry/data_secrecy/test_data_secrecy_logic.py

```diff
@@ -40,3 +54,272 @@ def test_bit_flag_disabled(self):
+
+    @patch("sentry.data_secrecy.data_secrecy_logic.data_access_grant_exists")
+    def test_should_allow_superuser_access_v2_no_active_grant(self, mock_grant_exists):
+        with self.settings(SENTRY_SELF_HOSTED=False):
+            with self.feature("organizations:data-secrecy-v2"):
+                mock_grant_exists.return_value = False
+                assert should_allow_superuser_access_v2(self.organization) is False
+                assert should_allow_superuser_access_v2(self.rpc_context) is False
+                mock_grant_exists.assert_called_with(self.organization.id)
+
+    def test_get_cached_grant_status_cache_miss(self):
+        # No cache entry exists
+        result = _get_cached_grant_status(self.organization.id)
+        assert result is None
+
+    def test_get_cached_grant_status_negative_cache_hit(self):
+        # Set negative cache value
+        cache_key = CACHE_KEY_PATTERN.format(organization_id=self.organization.id)
+        cache.set(cache_key, NEGATIVE_CACHE_VALUE, timeout=300)
+
+        result = _get_cached_grant_status(self.organization.id)
+        assert result is False
+
+    def test_get_cached_grant_status_valid_cached_data(self):
+        # Set valid cached data that hasn't expired
+        future_time = timezone.now() + timedelta(hours=1)
+        cached_data = RpcEffectiveGrantStatus(
+            organization_id=self.organization.id,
+            access_start=timezone.now(),
+            access_end=future_time,
+        ).dict()
+        cache_key = CACHE_KEY_PATTERN.format(organization_id=self.organization.id)
+        cache.set(cache_key, cached_data, timeout=300)
+
+        result = _get_cached_grant_status(self.organization.id)
+        assert result is True
+
+    def test_get_cached_grant_status_expired_cached_data(self):
+        # Set cached data that has logically expired
+        past_time = timezone.now() - timedelta(hours=1)
+        cached_data = RpcEffectiveGrantStatus(
+            organization_id=self.organization.id,
+            access_start=timezone.now() - timedelta(hours=2),
+            access_end=past_time,
+        ).dict()
+        cache_key = CACHE_KEY_PATTERN.format(organization_id=self.organization.id)
+        cache.set(cache_key, cached_data, timeout=300)
+
+        result = _get_cached_grant_status(self.organization.id)
+        assert result is None
+        # Verify cache was deleted
+        assert cache.get(cache_key) is None
+
+    # Tests for data_access_grant_exists
+    def test_data_access_grant_exists_cache_hit_positive(self):
+        # Set valid cached data
+        future_time = timezone.now() + timedelta(hours=1)
+        cached_data = RpcEffectiveGrantStatus(
+            organization_id=self.organization.id,
+            access_start=timezone.now(),
+            access_end=future_time,
+        ).dict()
+        cache_key = CACHE_KEY_PATTERN.format(organization_id=self.organization.id)
+        cache.set(cache_key, cached_data, timeout=300)
```

Comment:
I almost wonder if it makes sense to wrap the cache in its own class with its own methods to do this. It could make testing easier + less coupled overall.